File: frontend/src/api/research.ts
import { apiConfig, request, openEventStream } from "./client";
import { buildMarketIntelligenceScript, playScript, type RunHandle } from "@/mock/engine";
import { RESEARCH_RESULTS } from "@/mock/data/researchResults";
import type { ExecutionEvent } from "@/types/events";
import type { ResearchResult, ResearchRunSummary } from "@/types/research";

export interface StartRunResponse {
  runId: string;
}

let mockHandle: RunHandle | null = null;

/**
 * Domain service for the Market Intelligence engine.
 *
 * Endpoint shape this mirrors on the Python backend:
 *   POST /research/run                 -> { runId }
 *   GET  /research/runs/:id            -> ResearchRunSummary
 *   GET  /research/runs/:id/events     -> WebSocket stream of ExecutionEvent
 *   GET  /research/runs/:id/results    -> ResearchResult[]
 */
export const researchApi = {
  async startRun(): Promise<StartRunResponse> {
    if (apiConfig.useMocks) {
      return { runId: `run-mi-${Date.now()}` };
    }
    return request<StartRunResponse>("/research/run", { method: "POST" });
  },

  subscribeToRunEvents(runId: string, onEvent: (event: ExecutionEvent) => void): () => void {
    if (apiConfig.useMocks) {
      mockHandle?.cancel();
      const steps = buildMarketIntelligenceScript(RESEARCH_RESULTS);
      mockHandle = playScript("market_intelligence", runId, steps, onEvent);
      return () => mockHandle?.cancel();
    }
    return openEventStream(`/research/runs/${runId}/events`, (raw) => onEvent(JSON.parse(raw) as ExecutionEvent));
  },

  async getResults(): Promise<ResearchResult[]> {
    if (apiConfig.useMocks) return RESEARCH_RESULTS;
    return request<ResearchResult[]>(`/research/results`);
  },

  async getRunSummary(runId: string): Promise<ResearchRunSummary> {
    if (apiConfig.useMocks) {
      return { articlesScanned: 1284, eventsDetected: 143, candidatesSelected: RESEARCH_RESULTS.length, runDurationSec: 62 };
    }
    return request<ResearchRunSummary>(`/research/runs/${runId}`);
  },
};

File: backend/app/api/routes/research.py
import uuid

from fastapi import APIRouter, BackgroundTasks, Depends, WebSocket, WebSocketDisconnect
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.core.database import SessionLocal, get_db
from app.models.result import ResearchResult as ResearchResultModel
from app.models.run import EngineRun
from app.pipelines.research import run_market_intelligence
from app.schemas.research import ResearchResult, ResearchRunSummary, StartRunResponse
from app.ws.manager import run_event_broker

router = APIRouter(prefix="/research", tags=["research"])


@router.post("/run", response_model=StartRunResponse)
def start_run(background_tasks: BackgroundTasks, db: Session = Depends(get_db)) -> StartRunResponse:
    """Create a new Market Intelligence run and execute it in the background."""
    run_id = f"run-mi-{uuid.uuid4().hex[:12]}"
    db.add(EngineRun(id=run_id, engine="market_intelligence", status="PENDING"))
    db.commit()
    background_tasks.add_task(run_market_intelligence, run_id, SessionLocal)
    return StartRunResponse(runId=run_id)


@router.get("/runs/{run_id}", response_model=ResearchRunSummary)
def get_run_summary(run_id: str, db: Session = Depends(get_db)) -> ResearchRunSummary:
    """Return summary stats for a Market Intelligence run."""
    result_count = db.execute(
        select(ResearchResultModel).where(ResearchResultModel.run_id == run_id)
    ).scalars().all()
    run = db.get(EngineRun, run_id)
    duration = 0.0
    if run and run.started_at and run.finished_at:
        duration = (run.finished_at - run.started_at).total_seconds()
    return ResearchRunSummary(
        articlesScanned=0, eventsDetected=0, candidatesSelected=len(result_count), runDurationSec=duration,
    )


@router.get("/results", response_model=list[ResearchResult])
def get_results(db: Session = Depends(get_db)) -> list[dict]:
    """Return results from the most recently completed Market Intelligence run."""
    latest_run_id = db.execute(
        select(ResearchResultModel.run_id).order_by(ResearchResultModel.created_at.desc()).limit(1)
    ).scalar_one_or_none()
    if latest_run_id is None:
        return []
    rows = db.execute(
        select(ResearchResultModel).where(ResearchResultModel.run_id == latest_run_id)
    ).scalars().all()
    return [row.payload for row in rows]


@router.websocket("/runs/{run_id}/events")
async def run_events(websocket: WebSocket, run_id: str) -> None:
    """Stream ExecutionEvents for a run as they're published."""
    await run_event_broker.connect(run_id, websocket)
    try:
        while True:
            await websocket.receive_text()
    except WebSocketDisconnect:
        run_event_broker.disconnect(run_id, websocket)
